### PR TITLE
Make requests decode content

### LINKED ISSUE #3269: HfFileSystemStreamFile.read() does not handle Content-Encoding: gzip
### Describe the bug

Sometimes responses are returned from the server as gzip encoded:
```
{'Content-Type': 'text/plain; charset=utf-8', 'Transfer-Encoding': 'chunked', 'Connection': 'keep-alive', 'Date': 'Fri, 25 Jul 2025 16:51:10 GMT', 'Content-Security-Policy': "default-src 'none'; sandbox", 'Content-Disposition': 'inline; filename*=UTF-8\'\'math_qa.py; filename="math_qa.py";', 'X-Powered-By': 'huggingface-moon', 'cross-origin-opener-policy': 'same-origin', 'Referrer-Policy': 'strict-origin-when-cross-origin', 'X-Request-Id': 'Root=1-6883b5fe-41f5f016508b317149da7784;a76a214e-6e20-4ff6-8111-bc889e55b574', 'Access-Control-Allow-Origin': 'https://huggingface.co', 'Access-Control-Expose-Headers': 'X-Repo-Commit,X-Request-Id,X-Error-Code,X-Error-Message,X-Total-Count,ETag,Link,Accept-Ranges,Content-Range,X-Linked-Size,X-Linked-ETag,X-Xet-Hash', 'Set-Cookie': 'token=; Path=/; Expires=Thu, 01 Jan 1970 00:00:00 GMT; Secure; SameSite=None, token=; Domain=huggingface.co; Path=/; Expires=Thu, 01 Jan 1970 00:00:00 GMT; Secure; SameSite=Lax', 'X-Repo-Commit': 'c4f1cc784c04c4957b50c97858f23893b633eea6', 'Content-Encoding': 'gzip', 'ETag': 'W/"c7ec157d2c79e30553b7a65a9549277b34ad7d17"', 'Vary': 'Origin,accept-encoding', 'X-Cache': 'Hit from cloudfront', 'Via': '1.1 b9ba4bda57a72c60657b278a2341fc54.cloudfront.net (CloudFront)', 'X-Amz-Cf-Pop': 'CPH50-C1', 'X-Amz-Cf-Id': '8UQpDQ7v-q8ZEguQ5gqSrr6z_BK5IfCsYT3FoX4H6CSVZ_m3M3Vz8g==', 'Age': '233163'}
```

I suspect Cloudfront might be the cause.

In this case the `self.response.raw.read()` in [HfFileSystemStreamFile.read()](https://github.com/huggingface/huggingface_hub/blob/main/src/huggingface_hub/hf_file_system.py#L1072) will return the gzipped data instead of the actual data. This will result in an error like:
```
Failed to read the dataset script at datasets/math_qa/math_qa.py: 'utf-8' codec can't decode byte 0x8b in position 1: invalid start byte
```
in consumers. (If you google this error, you will see that others have encountered the same error and given up. Perhaps because it went away after some time, when Cloudfront changed it's cache.)

The easy fix is to add `self.response.raw.decode_content = True` just before `self.response.raw.read()`

### Reproduction

```
    fs = HfFileSystem()
    with fs.open(f"datasets/math_qa/math_qa.py", "r", encoding="utf-8") as f:
        print(f.read())
```

but note that this might vary depending on which exact cache Cloudfront edge server handles the request and what is has in its cache.

### Logs

```shell

```

### System info

```shell
- huggingface_hub version: 0.25.1
- Platform: Linux-6.14.6-061406-generic-x86_64-with-glibc2.41
- Python version: 3.12.10
- Running in iPython ?: No
- Running in notebook ?: No
- Running in Google Colab ?: No
- Running in Google Colab Enterprise ?: No
- Token path ?: /home/faber/.cache/huggingface/token
- Has saved token ?: False
- Configured git credential helpers: 
- FastAI: N/A
- Tensorflow: N/A
- Torch: 2.9.0.dev20250727+cu128
- Jinja2: 3.1.6
- Graphviz: N/A
- keras: N/A
- Pydot: N/A
- Pillow: 11.3.0
- hf_transfer: N/A
- gradio: N/A
- tensorboard: N/A
- numpy: 1.26.4
- pydantic: N/A
- aiohttp: 3.10.8
- ENDPOINT: https://huggingface.co
- HF_HUB_CACHE: /home/faber/.cache/huggingface/hub
- HF_ASSETS_CACHE: /home/faber/.cache/huggingface/assets
- HF_TOKEN_PATH: /home/faber/.cache/huggingface/token
- HF_HUB_OFFLINE: False
- HF_HUB_DISABLE_TELEMETRY: False
- HF_HUB_DISABLE_PROGRESS_BARS: None
- HF_HUB_DISABLE_SYMLINKS_WARNING: False
- HF_HUB_DISABLE_EXPERIMENTAL_WARNING: False
- HF_HUB_DISABLE_IMPLICIT_TOKEN: False
- HF_HUB_ENABLE_HF_TRANSFER: False
- HF_HUB_ETAG_TIMEOUT: 10
- HF_HUB_DOWNLOAD_TIMEOUT: 10
```

### PR DESCRIPTION
**Pull request by rasmusfaber**
If the response is gzipped, we should decode the content.

Fixes #3269

### CODE REVIEW

**Comment by Wauplin** [comment]
Hey @rasmusfaber, thanks for reporting https://github.com/huggingface/huggingface_hub/issues/3269 and opening this PR. I've just tried your patch locally and it doesn't seem to work for me. I used your repro example from the issue description:

```python
Python 3.10.12 (main, May 27 2025, 17:12:29) [GCC 11.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import requests
>>> requests.__version__
'2.32.3'
>>> import huggingface_hub 
>>> huggingface_hub.__version__
'0.34.0'
>>> from huggingface_hub import HfFileSystem
>>> fs = HfFileSystem()
>>> with fs.open(f"datasets/math_qa/math_qa.py", "r", encoding="utf-8") as f:
...     print(f.read())
... 
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
  File "/usr/lib/python3.10/codecs.py", line 322, in decode
    (result, consumed) = self._buffer_decode(data, self.errors, final)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xb2 in position 1: invalid start byte
```

**Comment by rasmusfaber** [comment]
> Hey @rasmusfaber, thanks for reporting #3269 and opening this PR. I've just tried your patch locally and it doesn't seem to work for me. I used your repro example from the issue description:

Hmm, strange. I will investigate.

**Comment by rasmusfaber** [comment]
> Hey @rasmusfaber, thanks for reporting #3269 and opening this PR. I've just tried your patch locally and it doesn't seem to work for me. I used your repro example from the issue description:

Sorry, can you try again? I was a little too quick to submit the fix. (`decode_content` is on `response.raw`, not on `response`).

**Comment by Wauplin** [comment]
Hmm, still not working (same error) but it might be on my side. I've put breakpoints just before `self.response.raw.decode_content = True` and it doesn't seem to be called at all.

**Comment by rasmusfaber** [comment]
> Hmm, still not working (same error) but it might be on my side. I've put breakpoints just before `self.response.raw.decode_content = True` and it doesn't seem to be called at all.

That sounds strange. 
`f.read()` should call [`HfFileSystemFile.read()`](https://github.com/huggingface/huggingface_hub/blob/53798919d5cb5486264d23ae8d8984400b3eadad/src/huggingface_hub/hf_file_system.py#L1004), which should call [`HfFileSystemStreamFile.read()`](https://github.com/rasmusfaber/huggingface_hub/blob/53798919d5cb5486264d23ae8d8984400b3eadad/src/huggingface_hub/hf_file_system.py#L1013).

I even tried to downgrade to Python 3.10.12 and it works as expected there as well.

Just to verify: are you running with the Git version of huggingface_hub (i.e. `pip install -e ".[dev]"` or equivalent)? I note that `huggingface_hub.__version__` is `0.35.0.dev0` for me and not `0.34.0` as you wrote above.

File: src/huggingface_hub/hf_file_system.py [hunk 1]
```diff
@@ -1069,6 +1069,7 @@ def read(self, length: int = -1):
             )
             hf_raise_for_status(self.response)
         try:
+            self.response.raw.decode_content = True
```
Comment (Wauplin) [inline src/huggingface_hub/hf_file_system.py:1072]: can you also set this line on [L1095](https://github.com/huggingface/huggingface_hub/pull/3271/files#diff-9c58c269bb02cc610db29c7d3066a9a02cc7430fc4175fe917b9c53abac72d3dR1095) as well? we have a second read call where it would be beneficial

Comment (Wauplin) [inline src/huggingface_hub/hf_file_system.py:1072]: can you also add a test in [test_hf_file_system.py](https://github.com/huggingface/huggingface_hub/blob/main/tests/test_hf_file_system.py) with your repro script


```py
    fs = HfFileSystem()
    with fs.open(f"datasets/math_qa/math_qa.py", "r", encoding="utf-8") as f:
        print(f.read())
```

that checks it runs without failure?

**Review by Wauplin — COMMENTED** [review]
Thanks @rasmusfaber! I'll never know what happened but a good old "delete and recreate env from scratch" solved it :smile: 

I've left 2 comments if you can address them before we get this merged?

File: src/huggingface_hub/hf_file_system.py (hunk 1 above)
Comment (rasmusfaber) [reply src/huggingface_hub/hf_file_system.py:1072]: Done.

Comment (rasmusfaber) [reply src/huggingface_hub/hf_file_system.py:1072]: I have added it. I think it is a bit dirty that it runs against production, but I don't know how I would ensure that the staging server returned a gzipped file.

Comment (Wauplin) [reply src/huggingface_hub/hf_file_system.py:1072]: Thanks! I think it's fine to run against production for such edge case test. It's not the only test where we do require production environment and so far it has been working correctly. As long as it's "read-only" it's perfectly ok. The load induced by the CI is neglictable compared to the Hub traffic in general (and worse case, it'll get rate limited before getting problematic)

**Review by Wauplin — APPROVED** [review]
Thanks a lot @rasmusfaber! Everything looks good to me. I'll merge as soon as CI is green :)

**Comment by Wauplin** [comment]
@bot /style

File: tests/test_hf_file_system.py [hunk 2]
```diff
@@ -612,3 +612,13 @@ def test_exists_after_repo_deletion():
...
     # Verify that the repo no longer exists.
     assert not hffs.exists(repo_id, refresh=True)
+
+
+@with_production_testing
+def test_hf_file_system_file_can_handle_gzipped_file():
+    """Test that HfFileSystemStreamFile.read() can handle gzipped files."""
+    fs = HfFileSystem()
+    # As of July 2025, the math_qa.py file is gzipped when queried from production:
+    with fs.open("datasets/math_qa/math_qa.py", "r", encoding="utf-8") as f:
```
Comment (Wauplin) [inline tests/test_hf_file_system.py]:
```suggestion
    with fs.open("datasets/allenai/math_qa/math_qa.py", "r", encoding="utf-8") as f:
```
